<launch>
    <arg name="rviz" default="true" />
    <arg name="scanPeriod" default="0.1" />
    <!-- Point clouds in Kitti odometry dataset are undistorted -->
    <arg name="pointUndistorted" default="true" />
    <!-- Kitti odometry dataset uses Velodyne HDL-64E -->
    <arg name="lidar" default="HDL-64E" />
    <arg name="publishMetrics" default="true" />
    <arg name="listenOdometry" default="true" />
    <!-- Set to false when not visualizing the output -->
    <arg name="publishFullPointCloud" default="true" />

    <node pkg="loam_velodyne" type="multiScanRegistration"
          name="multiScanRegistration" output="screen">
        <param name="systemDelay" value="0" />
        <param name="lidar" value="$(arg lidar)" />
        <param name="scanPeriod" value="$(arg scanPeriod)" />
        <param name="publishFullPointCloud"
               value="$(arg publishFullPointCloud)" />
        <param name="publishMetrics" value="$(arg publishMetrics)" />
        <remap from="/multi_scan_points" to="/velodyne_points" />
    </node>

    <node pkg="loam_velodyne" type="laserOdometry"
          name="laserOdometry" output="screen" respawn="true">
        <param name="scanPeriod" value="$(arg scanPeriod)" />
        <param name="pointUndistorted" value="$(arg pointUndistorted)" />
        <param name="publishFullPointCloud"
               value="$(arg publishFullPointCloud)" />
        <param name="publishMetrics" value="$(arg publishMetrics)" />
    </node>

    <node pkg="loam_velodyne" type="laserMapping"
          name="laserMapping" output="screen">
        <param name="scanPeriod" value="$(arg scanPeriod)" />
        <param name="publishFullPointCloud"
               value="$(arg publishFullPointCloud)" />
        <param name="publishMetrics" value="$(arg publishMetrics)" />
    </node>

    <node pkg="loam_velodyne" type="transformMaintenance"
          name="transformMaintenance" output="screen">
        <param name="publishMetrics" value="$(arg publishMetrics)" />
    </node>

    <group if="$(arg listenOdometry)">
        <node pkg="loam_velodyne" type="odometryListener"
              name="odometryListener" output="screen" />
    </group>

    <group if="$(arg publishMetrics)">
        <node pkg="loam_velodyne" type="metricsListener"
              name="metricsListener" output="screen" />
    </group>

    <group if="$(arg rviz)">
        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz"
              args="-d $(find loam_velodyne)/rviz_cfg/loam_velodyne.rviz" />
    </group>
</launch>
